### PR TITLE
Example how to read a shapefile and convert it to a geo-crate data-structure

### PR DESCRIPTION
**Pull request by michelk**

### CODE REVIEW

File: examples/read-polygon-shapefile-and-convert-to-geo-types/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+name = "read-polygon-shapefile-and-convert-to-geo-types"
+version = "0.1.0"
+authors = ["Michel Kuhlmann <michel.kuhlmann@tkconsult.ch>"]
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+shapefile = {version = "0.3.0", features = ["geo-types"]}
+dbase = "0.2.0"
```
Comment (tmontaigu) [inline examples/read-polygon-shapefile-and-convert-to-geo-types/Cargo.toml:11]: You actually don't need to add dbase as a dependency as its alreeady a shapefile dependency.
Remove it and in the main.rs add `use shapefile::dbase`

File: examples/read-polygon-shapefile-and-convert-to-geo-types/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            .expect("Could not open polygon-shapefile");
+    let points =
+        shapefile::read_as::<_, shapefile::Point, dbase::Record>("./tests/data/points.shp")
+            .expect("Could not open point shapefiles");
+    for (polygon, polygon_record) in polygons {
+        let geo_polygon: geo::MultiPolygon<f64> = polygon.into();
+        for (point, point_record) in points.iter() {
+            let geo_point: geo::Point<f64> = point.clone().into();
+            if geo_polygon.contains(&geo_point) {
+                let point_id = match point_record
```
Comment (tmontaigu) [inline examples/read-polygon-shapefile-and-convert-to-geo-types/src/main.rs:14]: For both of the `id` I find something like this to be cleaner:

```Rust
let point_id = match point_record.get("id") {
    Some(dbase::FieldValue::Numeric(Some(x))) => x,
    Some(_) => panic!("Expected 'id' to be a numeric"),
    None => panic!("Field 'id' is not within point-dataset"),
};
```

File: examples/read-polygon-shapefile-and-convert-to-geo-types/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+                    .expect("Field 'id' is not within polygon-dataset")
+                {
+                    dbase::FieldValue::Numeric(Some(x)) => x,
+                    _ => panic!("Could not parse id-field"),
+                };
+                println!(
+                    "Point with id {} is within polygon with id {}",
+                    point_id, polygon_id
+                );
+            } else {
```
Comment (tmontaigu) [inline examples/read-polygon-shapefile-and-convert-to-geo-types/src/main.rs:32]: That else has no purpose, you should remove it

**Review by tmontaigu — CHANGES_REQUESTED** [review]

**Comment by tmontaigu** [comment]
Also, I think the name of the folder is too long, it should be shorter, like "geotype-example" or "geotype-integration" or something else but shorter.

**Comment by michelk** [comment]
Thanks for the suggestions; I just applied them.

**Comment by tmontaigu** [comment]
Thanks for the example

**Comment by michelk** [comment]
Thanks. 

I have some problems to write another example: See [README](https://github.com/michelk/shapefile-rs/blob/read-example/examples/geotype-example/README.md) for a description and some [beginner code](https://github.com/michelk/shapefile-rs/blob/read-example/examples/geotype-example/src/bin/geotype-csv-example.rs).

I just have a knot in the brain, how to do the merge of csv::StringRecord and dbase::Record. Maybe you have some pointers? Thanks a lot.

**Comment by tmontaigu** [comment]
StringRecord seems to be a Collection of Strings so you will need to convert the content of the FieldValue to String.

**Comment by michelk** [comment]
Ok thanks. Now I have a working prototype. I added some information to the  [README](https://github.com/michelk/shapefile-rs/blob/read-example/examples/geotype-example/README.md) and the [ code](https://github.com/michelk/shapefile-rs/blob/read-example/examples/geotype-example/src/bin/geotype-csv-example.rs).

Would you consider to accept a PR with the new example?

**Comment by michelk** [comment]
Additionally, I had to add the Clone trait to dbase::Record, to get the example working.

**Comment by tmontaigu** [comment]
Yeah dbase::Record should derive Clone.

As for the example, i'm sure yet, I feel like this example is maybe too verbose, and stuff that is specific to the shapefile/dbase is too 'buried'.

One comment that I have about the code tho, is that the I think there is more `.clone()` that should be needed

**Pull request merged**